### PR TITLE
[Merged by Bors] - feat: ordered monoid instances for lexicographic order on Prod, Pi, Finsupp and DFinsupp

### PR DESCRIPTION
**Pull request by alreadydone**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Order/Monoid/Prod.lean [hunk 1]
```diff
@@ -41,4 +42,30 @@ instance [CanonicallyOrderedMonoid α] [CanonicallyOrderedMonoid β] :
     (inferInstance : ExistsMulOfLE _) with
       le_self_mul := fun _ _ ↦ ⟨le_self_mul, le_self_mul⟩ }
 
+namespace Lex
+
+@[to_additive]
+instance orderedCommMonoid [OrderedCommMonoid α]
+    [CovariantClass α α (· * ·) (· < ·)] [OrderedCommMonoid β] :
+    OrderedCommMonoid (α ×ₗ β) where
+  mul_le_mul_left := fun x y hxy z => ((le_iff _ _).1 hxy).elim
```
Comment (eric-wieser) [inline Mathlib/Algebra/Order/Monoid/Prod.lean:51]: Something like
```suggestion
  mul_le_mul_left x y hxy z := ((le_iff _ _).1 hxy).elim
```
should work, here and elsewhere. If the arguments are implicit, you'll need `{x y} hxy z` or similar instead.

Comment (alreadydone) [reply Mathlib/Algebra/Order/Monoid/Prod.lean]: Thanks, done! Wondering whether it can be made to work with `{ }`, not just `where`.

**Review by eric-wieser — APPROVED** [review]
Thanks! I pushed some changes to use `where` in more places, please revert if you dislike them.

bors d+

**Comment by alreadydone** [comment]
Thanks! Learnt a new syntax :)

bors r+

File: Mathlib/Data/DFinsupp/Lex.lean [hunk 2]
```diff
@@ -188,4 +188,34 @@ end Right
...
+
+instance Lex.orderedAddCommGroup [∀ i, OrderedAddCommGroup (α i)] :
+    OrderedAddCommGroup (Lex (Π₀ i, α i)) where
+  add_le_add_left _ _ := add_le_add_left
+
+instance Lex.linearOrderedCancelAddCommMonoid
+    [∀ i, LinearOrderedCancelAddCommMonoid (α i)] :
+    LinearOrderedCancelAddCommMonoid (Lex (Π₀ i, α i)) where
+  __ := (inferInstance : LinearOrder (Lex (Π₀ i, α i)))
+  __ := (inferInstance : OrderedCancelAddCommMonoid (Lex (Π₀ i, α i)))
```
Comment (alreadydone) [inline Mathlib/Data/DFinsupp/Lex.lean:212]:
```suggestion
  __ : LinearOrder (Lex (Π₀ i, α i)) := inferInstance
  __ : OrderedCancelAddCommMonoid (Lex (Π₀ i, α i)) := inferInstance
```
also works nicely

Comment (eric-wieser) [reply Mathlib/Data/DFinsupp/Lex.lean:212]: That's a neat trick!

**Pull request closed without merging**